### DISCUSSION TITLE
The dynamic supports Interface

### DISCUSSION THREAD

**Discussion by Flithor** [dotnet/csharplang#8915] [Language Ideas]
The idea is actually a kind of syntactic sugar:
As we all know, the `dynamic` keyword just skip the compilation check and compiles at runtime.
But dynamic cannot enjoy the convenience of Intellisense.
This proposal could bring some benefits.

Example:
```csharp
public dynamic interface MyDynamicInterface
{
    public int IntegerNumber { get; set; }
}

public void TestMethod(MyDynamicInterface arg, ThirdClass a, DynamicType b)
{
    MyDynamicInterface ia = a;
    MyDynamicInterface ib = b;
    // with Intellisense auto complete
    Console.WriteLine(arg.IntegerNumber);
    Console.WriteLine(ia.IntegerNumber);
    Console.WriteLine(ib.IntegerNumber);
}
```
Because it is syntactic sugar, it actually compiles to:
```csharp
public void TestMethod(dynamic arg, MySomeClass a, DynamicClass b)
{
    dynamic ia = a;
    dynamic ib = b;
    Console.WriteLine(arg.IntegerNumber);
    Console.WriteLine(ia.IntegerNumber);
    Console.WriteLine(ib.IntegerNumber);
}
```
`MyDynamicInterface` will not compile, so `typeof(MyDynamicInterface)` is invalid.

This proposal allows dynamic types to be treated as "strongly typed" with concrete definitions at compile time, and use same code on the classes that have no inheritance relationship but have the same members.

**Comment by colejohnson66** [comment]
I hate to play the broken record here, but this could be achieved with attributes and an analyzer:

```csharp
[DynamicInterface]
public interface MyDynamicInterface
{
    public int IntegerNumber { get; set; }
}

public void TestMethod(
    [DynamicArgument(typeof(MyDynamicInterface))] dynamic arg,
    [DynamicArgument(typeof(MyDynamicInterface))] dynamic a,
    [DynamicArgument(typeof(MyDynamicInterface))] dymamic b)
{
    Console.WriteLine(arg.IntegerNumber);
    Console.WriteLine(a.IntegerNumber);
    Console.WriteLine(b.IntegerNumber);
}
```

The analyzer would give an error if anyone actually implements the interface, and would give warnings/errors about accessing methods/properties that are not on the expected type. It wouldn't be as good as true IntelliSense, but it would work.

**Comment by bbarry** [comment]
https://github.com/dotnet/csharplang/discussions/2381
